src/app/clusters/energy-calendar-server/energy-calendar-server.cpp

```diff
@@ -0,0 +1,304 @@
+    {
+       NextPeakPeriod = nullptr; // todo
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+bool EnergyCalendarContent::CheckPeriods(DataModel::DecodableList<Structs::CalendarPeriod::Type> periods)
+{
+    uint32_t date = 0;
+    auto period_iterator = periods.begin();
+    while (period_iterator.Next())
+    {
+        auto & period = period_iterator.GetValue();
+        if (period.StartDate)
+        {
+            if (period.StartDate < date)
+            {
+                return false;
+            }
+            date = period.StartDate;
+        }
+
+        auto days_iterator = period.Days.begin();
+        while (days_iterator.Next())
+        {
+            auto & day = days_iterator.GetValue();
+            if (!CheckDay(day))
+            {
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+bool EnergyCalendarContent::CheckSpecialDays(DataModel::DecodableList<Structs::DayStruct::Type> days)
+{
+    uint32_t date = 0;
+    while (days_iterator.Next())
+    auto days_iterator = days.begin();
+    {
+        auto & day = days_iterator.GetValue();
+        if (day.DaysOfWeek || day.CalendarID || !CheckDay(day) || !(day.date) || day.date > date)
+        {
+            return false;
+        }
+        date = day.date;
+    }
+
+    return true;
+}
+
+bool EnergyCalendarContent::CheckDay(Structs::DayStruct::Type &day)
+{
+    if (day.DaysOfWeek && day.Date) || (!day.DaysOfWeek && !day.Date))
+    {
+        return false;
+    }
+
+    if (size(day.Transition)== 0)
+    {
+        return false;
+    }
+
+    uint32_t time = 0;
+    auto transition_iterator day.Transitions.begin();
+    while (transition_iterator.Next())
+    {
+        auto & transition = transition_iterator.GetValue();
+        auto tr_time = transition.TransitionTime;
+        if (tr_time < time)
+        {
+            return false;
+        }
+        time = tr_time;
+    }
+
+    return true;
+}
+
+// AttributeAccessInterface
+CHIP_ERROR EnergyCalendarServer::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    switch (aPath.mAttributeId)
+    {
+    case CalendarID::Id:
+        return aEncoder.Encode(content[endpointIndex].CalendarID);
+    case Name::Id:
+        return aEncoder.Encode(content[endpointIndex].Name);
+    case ProviderID::Id:
+        return aEncoder.Encode(content[endpointIndex].ProviderID);
+    case EventID::Id:
+        return aEncoder.Encode(content[endpointIndex].EventID);
+    case StartDate::Id:
+        return aEncoder.Encode(content[endpointIndex].StartDate);
+    case TimeReference::Id:
+        return aEncoder.Encode(content[endpointIndex].TimeReference);
+    case CalendarPeriods::Id:
+        return aEncoder.Encode(content[endpointIndex].CalendarPeriods);
+    case SpecialDays::Id:
+        return aEncoder.Encode(content[endpointIndex].SpecialDays);
+    /* Date relative attributes */
+    case CurrentDay::Id:
+        UpdateDateRelativeAttributes();
```

Comment:
Hold on.  Shouldn't this get updated at some points and subscription reports sent?  Or is this meant to be marked "C" in the spec?  It's not marked that...

Same for all the other things dependent on `UpdateDateRelativeAttributes`.